### PR TITLE
Update activations for MKL-DNN

### PR DESCRIPTION
**Pull request by kbinias**
Updated activations for MKL-DNN after changes in PaddlePaddle activations.

### CODE REVIEW

**Comment by luotao1** [comment]
> after changes in PaddlePaddle activations.

What's the changes do you mean?

**Comment by kbinias** [comment]
Changes for register activation operators (look at changes in activation_op.cc file, e.g. macros to create REGISTER_ACTIVATION_OP_MAKER class XXXOpMaker, new macros for register operators e.g. _INPLACE_, etc.)

File: paddle/fluid/operators/activation_op.cc [hunk 1]
```diff
@@ -458,21 +458,22 @@ namespace ops = paddle::operators;
...
-  __macro(Tanh, tanh);                       \
   __macro(Ceil, ceil);                       \
   __macro(Floor, floor);                     \
-  __macro(Sqrt, sqrt);                       \
   __macro(SoftRelu, soft_relu);              \
   __macro(Relu6, relu6);                     \
   __macro(Reciprocal, reciprocal);           \
   __macro(HardSigmoid, hard_sigmoid);
 
+#define FOR_EACH_MKLDNN_INPLACE_OP_FUNCTOR(__macro) \
```
Comment (luotao1) [inline paddle/fluid/operators/activation_op.cc]: Why changes of `activation_op.cc`?
- Are the previous registration not worked? But I see that the unit-test `test_activation_mkldnn_op.py` works well.
- Current changes are not suitable for other devices. For example, if there is `amd_relu_op`, how should it be defined?

File: paddle/fluid/operators/activation_mkldnn_op.cc [hunk 2]
```diff
@@ -52,9 +52,11 @@ void eltwise_forward(const ExecContext &ctx, mkldnn::algorithm algorithm,
...
 
   // create memory primitives
-  auto src_memory =
+  auto src_memory = std::make_shared<mkldnn::memory>(
       mkldnn::memory({data_md, mkldnn_engine},
-                     static_cast<void *>(const_cast<float *>(src_data)));
+                     static_cast<void *>(const_cast<float *>(src_data))));
+  // save source memory to device context to be referred in backward path
+  dev_ctx.SetBlob("InputX@eltwise_pd", src_memory);
   auto dst_memory =
```
[the remark this answers is not in the document]
Comment (kbinias) [reply paddle/fluid/operators/activation_mkldnn_op.cc]: For example, in MKL-DNN sqrt_bwd has different implementation than PaddlePaddle:

// MKL-DNN
template <typename T>
T sqrt_bwd(T dd, T s) {
    return s > 0 ? dd / (2 * ::sqrtf(s)) : 0;
}

// PP
void operator()(Device d, X x, Out out, dOut dout, dX dx) const {
   const Out out_conj = Eigen::numext::conj(out);
   dx.device(d) = static_cast<T>(0.5) * dout / out_conj;
}

this the reason I need input data in eltwise_grad

[the remark this answers is not in the document]
Comment (tensor-tang) [reply paddle/fluid/operators/activation_mkldnn_op.cc]: I know why `src_memory` is needed, but my point is why not also save `dst_memory` to context to save time.

File: paddle/fluid/operators/activation_mkldnn_op.cc [hunk 3]
```diff
@@ -69,7 +71,7 @@ void eltwise_forward(const ExecContext &ctx, mkldnn::algorithm algorithm,
       forward_desc, mkldnn_engine);
   dev_ctx.SetBlob(key_eltwise_pd, forward_pd);
```
[the remark this answers is not in the document]
Comment (tensor-tang) [reply paddle/fluid/operators/activation_mkldnn_op.cc]: Actually, my point is I can only see you always create and set this pd to context every forward iteration, but no reuse it in next iteration. Could we avoid this recreation among every iterations to enhance performance?

[the remark this answers is not in the document]
Comment (kbinias) [reply paddle/fluid/operators/activation_mkldnn_op.cc]: Yes, of course. I will improve it in next commit.

[the remark this answers is not in the document]
Comment (kbinias) [reply paddle/fluid/operators/activation_mkldnn_op.cc]: I added suport for reusing memory buffers to MKL-DNN activations.

**Comment by kbinias** [comment]
Could you possibly continue your review ? I added suport for reusing memory buffers.

File: paddle/fluid/operators/activation_mkldnn_op.cc [hunk 4]
```diff
@@ -23,6 +24,18 @@ using paddle::framework::Tensor;
 using paddle::platform::MKLDNNDeviceContext;
 
 namespace {
+std::string gethash(const mkldnn::memory::dims &operand_dims,
+                    const mkldnn::algorithm algorithm) {
+  auto dim2str = [](const mkldnn::memory::dims &operand_dims) {
```
Comment (tensor-tang) [inline paddle/fluid/operators/activation_mkldnn_op.cc]: Is that possible two fc ops with same src dims, then would this be right?  Then would share same hash code?

Comment (kbinias) [reply paddle/fluid/operators/activation_mkldnn_op.cc]: Probably yes, but they will be reusable. I improved hash code for input data (key_src_data)

Comment (tensor-tang) [reply paddle/fluid/operators/activation_mkldnn_op.cc]: OK, good.

**Pull request merged**